IllegalArgumentException caused by local default datetime when using nullAsDefault property

` jdbc:ch:http://host:port?nullAsDefault=2`

```java
// ClickHouseDateTimeValue.class
static {
    DEFAULT = ClickHouseDateValue.DEFAULT.atStartOfDay();
    dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
}

// BinaryStreamUtils.class
public static void writeDateTime32(OutputStream output, LocalDateTime value, TimeZone tz) throws IOException {
        long time = tz != null && !tz.equals(ClickHouseValues.UTC_TIMEZONE) ? value.atZone(tz.toZoneId()).toEpochSecond() : value.toEpochSecond(ZoneOffset.UTC);
        writeUnsignedInt32(output, ClickHouseChecker.between(time, "DateTime", 0L, 4294967295000L));
 }
```

```
Caused by: java.lang.IllegalArgumentException: DateTime(-28800) should between 0 and 4294967295000 inclusive of both values
	at com.clickhouse.client.ClickHouseChecker.newException(ClickHouseChecker.java:19) ~[clickhouse-jdbc-0.3.2-patch10-http.jar:clickhouse-jdbc 0.3.2-patch10 (revision: aebad16)]
	at com.clickhouse.client.ClickHouseChecker.between(ClickHouseChecker.java:111) ~[clickhouse-jdbc-0.3.2-patch10-http.jar:clickhouse-jdbc 0.3.2-patch10 (revision: aebad16)]
	at com.clickhouse.client.data.BinaryStreamUtils.writeDateTime32(BinaryStreamUtils.java:1468) ~[clickhouse-jdbc-0.3.2-patch10-http.jar:clickhouse-jdbc 0.3.2-patch10 (revision: aebad16)]
	at com.clickhouse.client.data.BinaryStreamUtils.writeDateTime(BinaryStreamUtils.java:1433) ~[clickhouse-jdbc-0.3.2-patch10-http.jar:clickhouse-jdbc 0.3.2-patch10 (revision: aebad16)]
	at com.clickhouse.client.data.ClickHouseRowBinaryProcessor$MappedFunctions.lambda$buildMappingsForDataTypes$53(ClickHouseRowBinaryProcessor.java:299) ~[clickhouse-jdbc-0.3.2-patch10-http.jar:clickhouse-jdbc 0.3.2-patch10 (revision: aebad16)]
	at com.clickhouse.client.data.ClickHouseRowBinaryProcessor$MappedFunctions.serialize(ClickHouseRowBinaryProcessor.java:486) ~[clickhouse-jdbc-0.3.2-patch10-http.jar:clickhouse-jdbc 0.3.2-patch10 (revision: aebad16)]
	at com.clickhouse.jdbc.internal.InputBasedPreparedStatement.addBatch(InputBasedPreparedStatement.java:326) ~[clickhouse-jdbc-0.3.2-patch10-http.jar:clickhouse-jdbc 0.3.2-patch10 (revision: aebad16)]
	at com.clickhouse.jdbc.internal.InputBasedPreparedStatement.executeAny(InputBasedPreparedStatement.java:104) ~[clickhouse-jdbc-0.3.2-patch10-http.jar:clickhouse-jdbc 0.3.2-patch10 (revision: aebad16)]
	at com.clickhouse.jdbc.internal.InputBasedPreparedStatement.execute(InputBasedPreparedStatement.java:294) ~[clickhouse-jdbc-0.3.2-patch10-http.jar:clickhouse-jdbc 0.3.2-patch10 (revision: aebad16)]

```

--- a/clickhouse-client/src/main/java/com/clickhouse/client/data/ClickHouseDateTimeValue.java
+++ b/clickhouse-client/src/main/java/com/clickhouse/client/data/ClickHouseDateTimeValue.java
@@ -9,6 +9,7 @@ import java.time.LocalDate;
 import java.time.LocalDateTime;
 import java.time.LocalTime;
 import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
 import java.time.ZonedDateTime;
 import java.time.format.DateTimeFormatter;
 import java.util.TimeZone;
@@ -24,7 +25,8 @@ public class ClickHouseDateTimeValue extends ClickHouseObjectValue<LocalDateTime
     /**
      * Default value.
      */
-    public static final LocalDateTime DEFAULT = ClickHouseDateValue.DEFAULT.atStartOfDay();
+    public static final LocalDateTime DEFAULT = ClickHouseInstantValue.DEFAULT.atOffset(ZoneOffset.UTC)
+            .toLocalDateTime();
 
     static final DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");
 
@@ -111,11 +113,15 @@ public class ClickHouseDateTimeValue extends ClickHouseObjectValue<LocalDateTime
 
     private final int scale;
     private final TimeZone tz;
+    private final LocalDateTime defaultValue;
 
     protected ClickHouseDateTimeValue(LocalDateTime value, int scale, TimeZone tz) {
         super(value);
         this.scale = ClickHouseChecker.between(scale, ClickHouseValues.PARAM_SCALE, 0, 9);
         this.tz = tz != null ? tz : ClickHouseValues.UTC_TIMEZONE;
+        this.defaultValue = this.tz.equals(ClickHouseValues.UTC_TIMEZONE) ? DEFAULT
+                : ClickHouseOffsetDateTimeValue.DEFAULT.toZonedDateTime().withZoneSameInstant(this.tz.toZoneId())
+                        .toLocalDateTime();
     }
 
     public int getScale() {
@@ -237,7 +243,7 @@ public class ClickHouseDateTimeValue extends ClickHouseObjectValue<LocalDateTime
 
     @Override
     public ClickHouseDateTimeValue resetToDefault() {
-        set(DEFAULT);
+        set(defaultValue);
         return this;
     }
 

--- a/clickhouse-client/src/main/java/com/clickhouse/client/data/ClickHouseDateValue.java
+++ b/clickhouse-client/src/main/java/com/clickhouse/client/data/ClickHouseDateValue.java
@@ -19,7 +19,7 @@ public class ClickHouseDateValue extends ClickHouseObjectValue<LocalDate> {
     /**
      * Default date.
      */
-    public static final LocalDate DEFAULT = LocalDate.of(1970, 1, 1);
+    public static final LocalDate DEFAULT = ClickHouseOffsetDateTimeValue.DEFAULT.toLocalDate();
 
     /**
      * Create a new instance representing null value.

--- a/clickhouse-client/src/main/java/com/clickhouse/client/data/ClickHouseOffsetDateTimeValue.java
+++ b/clickhouse-client/src/main/java/com/clickhouse/client/data/ClickHouseOffsetDateTimeValue.java
@@ -87,11 +87,15 @@ public class ClickHouseOffsetDateTimeValue extends ClickHouseObjectValue<OffsetD
 
     private final int scale;
     private final TimeZone tz;
+    private final OffsetDateTime defaultValue;
 
     protected ClickHouseOffsetDateTimeValue(OffsetDateTime value, int scale, TimeZone tz) {
         super(value);
         this.scale = ClickHouseChecker.between(scale, ClickHouseValues.PARAM_SCALE, 0, 9);
         this.tz = tz == null || tz.equals(ClickHouseValues.UTC_TIMEZONE) ? ClickHouseValues.UTC_TIMEZONE : tz;
+        this.defaultValue = this.tz.equals(ClickHouseValues.UTC_TIMEZONE) ? DEFAULT
+                : ClickHouseInstantValue.DEFAULT
+                        .atOffset(tz.toZoneId().getRules().getOffset(ClickHouseInstantValue.DEFAULT));
     }
 
     public int getScale() {
@@ -221,9 +225,7 @@ public class ClickHouseOffsetDateTimeValue extends ClickHouseObjectValue<OffsetD
 
     @Override
     public ClickHouseOffsetDateTimeValue resetToDefault() {
-        set(tz == null ? DEFAULT
-                : ClickHouseInstantValue.DEFAULT
-                        .atOffset(tz.toZoneId().getRules().getOffset(ClickHouseInstantValue.DEFAULT)));
+        set(defaultValue);
         return this;
     }
 

--- a/clickhouse-jdbc/src/test/java/com/clickhouse/jdbc/ClickHousePreparedStatementTest.java
+++ b/clickhouse-jdbc/src/test/java/com/clickhouse/jdbc/ClickHousePreparedStatementTest.java
@@ -20,6 +20,7 @@ import java.time.Instant;
 import java.time.LocalDate;
 import java.time.LocalDateTime;
 import java.time.ZoneOffset;
+import java.util.Calendar;
 import java.util.Collections;
 import java.util.Properties;
 import java.util.TimeZone;
@@ -36,7 +37,9 @@ import com.clickhouse.client.ClickHousePipedOutputStream;
 import com.clickhouse.client.ClickHouseProtocol;
 import com.clickhouse.client.config.ClickHouseClientOption;
 import com.clickhouse.client.data.ClickHouseBitmap;
+import com.clickhouse.client.data.ClickHouseDateTimeValue;
 import com.clickhouse.client.data.ClickHouseExternalTable;
+import com.clickhouse.client.data.ClickHouseOffsetDateTimeValue;
 import com.clickhouse.jdbc.internal.InputBasedPreparedStatement;
 import com.clickhouse.jdbc.internal.SqlBasedPreparedStatement;
 
@@ -1475,6 +1478,39 @@ public class ClickHousePreparedStatementTest extends JdbcIntegrationTest {
                 Assert.assertEquals(rs.getObject(1), Integer.valueOf(5));
                 Assert.assertEquals(rs.getString(2), null);
                 Assert.assertTrue(rs.wasNull());
+                Assert.assertFalse(rs.next());
+            }
+        }
+    }
+
+    @Test(groups = "integration")
+    public void testInsertWithNullDateTime() throws Exception {
+        Properties props = new Properties();
+        props.setProperty(JdbcConfig.PROP_NULL_AS_DEFAULT, "2");
+        try (ClickHouseConnection conn = newConnection(props);
+                Statement s = conn.createStatement()) {
+            s.execute("drop table if exists test_insert_with_null_datetime; "
+                    + "CREATE TABLE test_insert_with_null_datetime(a Int32, "
+                    + "b01 DateTime32, b02 DateTime32('America/Los_Angeles'), "
+                    + "b11 DateTime32, b12 DateTime32('America/Los_Angeles'), "
+                    + "c01 DateTime64(3), c02 DateTime64(6, 'Asia/Shanghai'), "
+                    + "c11 DateTime64(3), c12 DateTime64(6, 'Asia/Shanghai')) ENGINE=Memory");
+            try (PreparedStatement ps = conn
+                    .prepareStatement("INSERT INTO test_insert_with_null_datetime values(?, ? ,? ,?,?)")) {
+                ps.setInt(1, 1);
+                ps.setObject(2, LocalDateTime.now());
+                ps.setObject(3, LocalDateTime.now());
+                ps.setTimestamp(4, null);
+                ps.setNull(5, Types.TIMESTAMP);
+                ps.setObject(6, LocalDateTime.now());
+                ps.setObject(7, LocalDateTime.now());
+                ps.setObject(8, null);
+                ps.setTimestamp(9, null, Calendar.getInstance());
+                ps.executeUpdate();
+            }
+
+            try (ResultSet rs = s.executeQuery("select * from test_insert_with_null_datetime order by a")) {
+                Assert.assertTrue(rs.next());
                 Assert.assertFalse(rs.next());
             }
         }

